File: 2019-01-30/helloworldbeam/src/main/java/Ex01.java
import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.TextIO;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.Create;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.ParDo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;

public class Ex01 {
    private static final Logger LOG = LoggerFactory.getLogger(Ex01.class);

    public static class MyFunction extends DoFn<Integer, String> {
        @ProcessElement
        public void processElement(ProcessContext context) {
            Integer value = context.element();

            LOG.info(String.format("value=%d", value));

            context.output(value.toString());
        }
    }

    public static void main(String args[]) {
        PipelineOptions options = PipelineOptionsFactory
                .fromArgs(args)
                .withValidation()
                .as(PipelineOptions.class);

        Pipeline p = Pipeline
                .create(options);

        p
                .apply(Create.of(Arrays.asList(1,2,3,4,5,6,7,8,9,10)))
                .apply(ParDo.of(new MyFunction()))
                .apply(TextIO
                            .write()
                            .to("gs://arquivei-curso-dataflow/ex01/"));

        p.run().waitUntilFinish();
    }
}

File: 2019-01-31/helloworldscio/src/main/scala/Ex01.scala
import com.spotify.scio.ContextAndArgs
import com.spotify.scio.values.SCollection.State
import org.apache.beam.sdk.state.{StateSpecs, ValueState}
import org.apache.beam.sdk.transforms.{DoFn, ParDo}
import org.apache.beam.sdk.transforms.DoFn.{ProcessElement, StateId}
import org.apache.beam.sdk.values.KV
import org.slf4j.LoggerFactory

object Ex01 {
  val logger = LoggerFactory.getLogger(this.getClass)

  class SumPerKey extends DoFn[KV[String, Double], KV[String, Double]] {
    @StateId("keySum") private val keySum = StateSpecs.value[java.lang.Double]()

    @ProcessElement
    def processElement(ctx: ProcessContext,
                       @StateId("keySum") elemKeySum: ValueState[java.lang.Double]) = {
      val kv = ctx.element()

      val key = kv.getKey
      val value = kv.getValue

      val currentSum = elemKeySum.read()
      logger.info(s"kv=${kv.toString} currentSum=$currentSum")

      val newSum = currentSum+value
      logger.info(s"kv=${kv.toString} newSum=$newSum")

      elemKeySum.write(newSum)

      ctx.output(kv)
    }
  }

  def main(args: Array[String]): Unit = {
    val (sc, _) = ContextAndArgs(args)

    sc.pubsubSubscription[String]("projects/arquivei-curso-dataflow/subscriptions/curso-dataflow")
      .withName("print all args").map { value =>
        logger.info(s"value=$value")
        val tokens = value.split("=")
        (tokens(0), tokens(1).toDouble)
      }
      .withGlobalWindow()
      .withName("sum values").applyPerKeyDoFn(new SumPerKey)

    sc.close().waitUntilFinish()
  }
}
